src/Tasks/Hash.cs

```diff
@@ -37,61 +35,120 @@ public class Hash : TaskExtension
                 {
-                    var concatenatedItemStringSize = ComputeStringSize(ItemsToHash);
+                    // Buffer in which bytes of the strings are to be stored until their number reachs the limit size.
+                    // Once the limit is reached, the sha1.TransformBlock is be run on all the bytes of this buffer.
+                    byte[] sha1Buffer = System.Buffers.ArrayPool<byte>.Shared.Rent(sha1BufferSize);
 
-                    var hashStringSize = sha1.HashSize;
+                    int maxItemStringSize = encoding.GetMaxByteCount(Math.Min(ComputeMaxItemSpecLength(ItemsToHash), maxInputChunkLength));
+                    byte[] bytesBuffer = System.Buffers.ArrayPool<byte>.Shared.Rent(maxItemStringSize);
 
-                    using (var stringBuilder = new ReuseableStringBuilder(Math.Max(concatenatedItemStringSize, hashStringSize)))
+                    int bufferLength = 0;
+                    for (int i = 0; i < ItemsToHash.Length; i++)
                     {
-                        foreach (var item in ItemsToHash)
+                        string itemSpec = IgnoreCase ? ItemsToHash[i].ItemSpec.ToUpperInvariant() : ItemsToHash[i].ItemSpec;
+
+                        // Slice the itemSpec string into chunks of reasonable size and add them to sha1 buffer.
+                        int itemSpecLength = itemSpec.Length;
+                        int itemSpecPosition = 0;
+                        while (itemSpecLength > 0)
                         {
-                            string itemSpec = item.ItemSpec;
-                            stringBuilder.Append(IgnoreCase ? itemSpec.ToUpperInvariant() : itemSpec);
-                            stringBuilder.Append(ItemSeparatorCharacter);
+                            int charsToProcess = Math.Min(itemSpecLength, maxInputChunkLength);
+                            int bytesNumber = encoding.GetBytes(itemSpec, itemSpecPosition, charsToProcess, bytesBuffer, 0);
+                            itemSpecPosition += charsToProcess;
+                            itemSpecLength -= charsToProcess;
+
+                            AddBytesToSha1Buffer(sha1, sha1Buffer, ref bufferLength, sha1BufferSize, bytesBuffer, bytesNumber);
                         }
 
-                        var hash = sha1.ComputeHash(Encoding.UTF8.GetBytes(stringBuilder.ToString()));
+                        AddBytesToSha1Buffer(sha1, sha1Buffer, ref bufferLength, sha1BufferSize, ItemSeparatorCharacterBytes, ItemSeparatorCharacterBytes.Length);
+                    }
 
-                        stringBuilder.Clear();
+                    sha1.TransformFinalBlock(sha1Buffer, 0, bufferLength);
 
-                        foreach (var b in hash)
+                    using (var stringBuilder = new ReuseableStringBuilder(sha1.HashSize))
+                    {
+                        foreach (var b in sha1.Hash)
                         {
                             stringBuilder.Append(b.ToString("x2"));
```

Comment:
Can AppendFormat with something like "{0:x2}" be used instead? Not sure.

Reply:
What's the difference?

Reply:
Avoids intermediate ToString allocation and uses writes directly to the internal stringbuilder buffer.

Reply:
current: value.ToString(fmt) -> Append(string)
proposed: AppendFormat(fmt, value)

The only downside is that you probably box the int, so allocation still occurs, but smaller than original.

Reply:
yeah, I suppose we can use it here. I will try it.

Reply:
hmm, I suppose I will not switch to `AppendFormat`: there is no function `AppendFormat` in ReuseableStringBuilder at the moment and I do not see improvement when I tried to switch in benchmark results. I do not think it is worth additional work here.